pkg/common/cns-lib/volume/listview.go

```diff
@@ -0,0 +1,348 @@
+package volume
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/property"
+	"github.com/vmware/govmomi/session"
+	"github.com/vmware/govmomi/view"
+	"github.com/vmware/govmomi/vim25/types"
+
+	cnsvsphere "sigs.k8s.io/vsphere-csi-driver/v2/pkg/common/cns-lib/vsphere"
+	"sigs.k8s.io/vsphere-csi-driver/v2/pkg/csi/service/logger"
+)
+
+const (
+	// in case of vc failures, we wait for a minute
+	// to not overwhelm the vc with continuous requests
+	waitForUpdatesRetry = 1 * time.Minute
+	// `info` is a defined property in a Task object
+	// we want to monitor the TaskInfo that includes the status of a task
+	infoPropertyName = "info"
+)
+
+// ListViewImpl is the struct used to manage a single listView instance.
+// params
+// sync.RWMutex: is required to safely handle concurrent map ops
+// virtualCenter: holds a reference to the global VC object
+// listView: holds the managed object used to monitor multiple concurrent VC tasks
+// context.Context: new context for the life of the listview object.
+// it's separate from the context set by CSI ops
+// taskMap: is an in-memory map to associate task details
+// to the channel created by the caller to receive the task result
+// shouldStopListening: in case of regular CSI operation, even after receiving a batch of updates,
+// we want to continue listening for subsequent updates.
+// in case of unit tests, we need to stop listening for the test to complete execution
+type ListViewImpl struct {
+	mu                  sync.RWMutex
```

Comment:
Can we have an individual mutex to protect govmomi, vcenter and taskMap.

Reply:
I've kept a mutex for the map. I realized we don't need a mutex at any other place in the code.